File: bits/2220-minimum-bit-flips-to-convert-number.py
import unittest

# https://leetcode.com/problems/minimum-bit-flips-to-convert-number/

# python3 -m unittest bits/2220-minimum-bit-flips-to-convert-number.py

class Solution(unittest.TestCase):
    # # Approach: Iterative
    # # Time: O(max bits)
    # # Space: O(1)
    # def minBitFlips(self, start: int, goal: int) -> int:
    #     flips = 0
    #     xor_result = start ^ goal
    #     while xor_result:
    #         flips += xor_result&1
    #         xor_result >>= 1
    #     return flips

    # # Approach: Recursive
    # # Time: O(max bits)
    # # Space: O(max bits)
    # def minBitFlips(self, start: int, goal: int) -> int:
    #     if not start and not goal:
    #         return 0
    #     flips = (start&1) ^ (goal&1)
    #     return flips + self.minBitFlips(start>>1, goal>>1)

    # Approach: Brian Kernighan’s Algorithm
    # Time: O(number of set bits)
    # Space: O(1)
    def minBitFlips(self, start: int, goal: int) -> int:
        # return bin(start^goal).count('1')
        # return (start^goal).bit_count()
        flips = 0
        xor_result = start ^ goal
        while xor_result:
            xor_result &= xor_result-1
            flips += 1
        return flips

    def test(self):
        for start, goal, expected in [
            (10, 7, 3),
            (3, 4, 3),
        ]:
            output = self.minBitFlips(start, goal)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: bits/2220-minimum-bit-flips-to-convert-number.go
package bits

// https://leetcode.com/problems/minimum-bit-flips-to-convert-number/

// Note: This question is the same as 461: Hamming Distance.

// Approach #3: Brian Kernighan’s Algorithm
// Time: O(number of set bits)
// Space: O(1)
func minBitFlips(start int, goal int) int {
	// return bits.OnesCount(uint(start ^ goal))
	var flips = 0
	var result = start ^ goal
	for result != 0 {
		result &= result - 1
		flips += 1
	}
	return flips
}

// // Approach #2: Iterative
// // Time: O(max bits)
// // Space: O(1)
// func minBitFlips(start int, goal int) int {
// 	var flips = 0
// 	var result = start ^ goal
// 	for result != 0 {
// 		flips += result & 1
// 		result >>= 1
// 	}
// 	return flips
// }

// // Approach #1: Recursive
// // Time: O(max bits)
// // Space: O(max bits)
// func minBitFlips(start int, goal int) int {
// 	if start == 0 && goal == 0 {
// 		return 0
// 	}
// 	var flips = (start & 1) ^ (goal & 1)
// 	return flips + minBitFlips(start>>1, goal>>1)
// }
